### PR TITLE
Issue 37: XA Store

### PR DESCRIPTION
**Pull request by lorban**
This PR is about the full XA store. Things that need to be reviewed with a bit more attention:

 - the couple of changes I had to make in api/core/impl
 - the asciidoc "integration"
 - the independency / integration of the "transactions" module
 - the config settings "look and feel"
 - everything else!

### CODE REVIEW

**Comment by lorban** [comment]
Things that have been left aside:
 - Integration with different transaction managers. Currently only BTM is supported and has been tested. Though, the XA mechnisms used should work with other impls too and the TX mgr lookup code has been made pluggable.
 - XAStore's putIfAbsent(), replace(2 args), replace(3 args), remove(2 args) because they currently are uneeded.
 - Non-strict XA, i.e.: using a JTA Synchronization to sync the XA store with a TM.

File: impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java [hunk 1]
```diff
@@ -58,20 +55,24 @@ public DefaultCopyProvider(DefaultCopyProviderConfiguration configuration) {
...
-    return createCopier(clazz, conf);
+    LOG.info("Copier for <{}> : {}", clazz.getName(), copier);
+    return copier;
   }
 
   private <T> Copier<T> createCopier(Class<T> clazz, DefaultCopierConfiguration<T> config) {
     String alias = (config != null ? null : clazz.getName());
     Copier<T> copier = (Copier<T>) newInstance(alias, config);
     if (copier == null) {
-      LOG.info("No registered copier found for for <{}>. Using the default Identity copier.", clazz.getName());
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java]: Might be interesting to keep that one as debug

File: transactions/src/docs/asciidoc/user/xa.adoc [hunk 2]
```diff
@@ -0,0 +1,207 @@
+include::common.adoc[]
```
Comment (ljacomet) [inline transactions/src/docs/asciidoc/user/xa.adoc]: Why is this file here and not in the docs module?

File: transactions/config/checkstyle-suppressions.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!DOCTYPE suppressions PUBLIC
+    "-//Puppy Crawl//DTD Suppressions 1.1//EN"
+    "http://www.puppycrawl.com/dtds/suppressions_1_1.dtd">
+
+<suppressions>
+  <suppress files="^((?!.*test[\\/]java[\\/]org[\\/]ehcache[\\/]transactions[\\/]xa[\\/]docs[\\/].*).)*$" checks="AvoidStaticImport"/>
```
Comment (ljacomet) [inline transactions/config/checkstyle-suppressions.xml]: Seems to indicate that you want samples for doc in a different package. Why?

**Comment by ljacomet** [comment]
Mostly made comments around the module setup and integration with the project.

On that front, it is missing integration in building the ehcache shaded jar. I am assuming we want that jar to contain transaction stuff from now on.

File: impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java (hunk 1 above)
Comment (lorban) [inline impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java]: That particular location wasn't ideal, so I've moved that INFO log a few lines above which makes the logger output less confusing.

File: transactions/config/checkstyle-suppressions.xml (hunk 3 above)
Comment (lorban) [inline transactions/config/checkstyle-suppressions.xml]: For no special reason. I wondered how independent the transactions modules has to be, do I bundled everything including the doc in it. I could move all that to the docs module if you believe this makes more sense.

File: transactions/src/docs/asciidoc/user/xa.adoc (hunk 2 above)
Comment (lorban) [inline transactions/src/docs/asciidoc/user/xa.adoc]: Same answer as below: I can move the XA doc to the docs module if you believe it makes more sense.

File: impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java (hunk 1 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/spi/copy/DefaultCopyProvider.java]: Right, no need to log twice, the other log entry indicates which copier is used always.

File: api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  /**
+   * Create an instance of RethrowingCacheAccessException.
+   * @param cause the cause RuntimeException that will be rethrown.
+   */
+  public RethrowingCacheAccessException(RuntimeException cause) {
+    super(cause);
+  }
+
+  @Override
+  public synchronized RuntimeException getCause() {
```
Comment (ljacomet) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: Why `synchronized`?

File: transactions/src/main/java/org/ehcache/transactions/xa/configuration/XAStoreConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import org.ehcache.transactions.xa.XAStore;
+
+/**
+ * @author Ludovic Orban
+ */
+public class XAStoreConfiguration implements ServiceConfiguration<XAStore.Provider> {
+
+  private final String uniqueXAResourceId;
+
+  //TODO: ideally, the uniqueXAResourceIdSuffix should be the cache alias, injected by the CM
```
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/configuration/XAStoreConfiguration.java]: Is that still a _TODO_? Cause it mentions something that no longer seem to exist

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java [hunk 6]
```diff
@@ -0,0 +1,908 @@
...
+
+        // check if the store is persistent, and get the PersistenceSpaceIdentifier if it is
+        LocalPersistenceService.PersistenceSpaceIdentifier persistenceSpaceId = null;
+        ResourcePools resourcePools = storeConfig.getResourcePools();
+        for (ResourceType resourceType : resourcePools.getResourceTypeSet()) {
+          ResourcePool pool = resourcePools.getPoolForResource(resourceType);
+          if (pool.isPersistent()) {
+            try {
+              LocalPersistenceService persistenceService = serviceProvider.getService(LocalPersistenceService.class);
+              persistenceSpaceId = persistenceService.getOrCreatePersistenceSpace(uniqueXAResourceId);
```
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: Are you sure you need a different storage _space_? That is one disconnected from the _space_ used by this specific cache?
Why don't you look up the persistenceSpaceId passed in the service configurations?
Then the data would end up in the same location as the rest of the data for the cache, and so if the cache is destroyed then this data is destroyed as well.

**Comment by ljacomet** [comment]
PR needs a rebase, for some reason it is based on the last real commit and not the merge commit.

File: api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java (hunk 4 above)
Comment (lorban) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: Because the overridden implementation is synchonized too.

File: transactions/src/main/java/org/ehcache/transactions/xa/configuration/XAStoreConfiguration.java (hunk 5 above)
Comment (lorban) [inline transactions/src/main/java/org/ehcache/transactions/xa/configuration/XAStoreConfiguration.java]: First: uniqueXAResourceIdSuffix should read uniqueXAResourceId, I'll fix that

Then, this can be discussed; we may still want the unique XAResource ID to default to the cache name at a later stage, or close that door forever.

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java (hunk 6 above)
Comment (lorban) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: Our understandings aren't aligned: mine is that this code creates a subspace into the cache's space. I'll investigate and if I'm wrong, I'll change this code accordingly.

**Comment by lorban** [comment]
test this please

File: api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java (hunk 4 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: What?

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java (hunk 6 above)
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: For that you need to invoke `org.ehcache.internal.persistence.DefaultLocalPersistenceService#createPersistenceContextWithin` which takes a persistenceSpaceId as first param and a name (the subfolder) as second.

File: api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java (hunk 4 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: So not `synchronised` in 1.6, well in 1.7 and 1.8 ... ok, good for me then

Comment (lorban) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: Still weird...

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java (hunk 6 above)
Comment (lorban) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: The createPersistenceContextWithin() call is done in DefaultJournalProvider. But you're right that there is no need to create a new persistence space, I can just lookup the space ID that is in the service configs instead and reuse the cache's space.

**Comment by lorban** [comment]
test this please

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+      final Transaction transaction = transactionManagerWrapper.getTransactionManager().getTransaction();
+      if (transaction == null) {
+        throw new XACacheAccessException(new XACacheException("Cannot access XA cache outside of XA transaction scope"));
+      }
+      EhcacheXAResource<K, V> xaResource = xaResources.get(transaction);
+      if (xaResource == null) {
+        xaResource = new EhcacheXAResource<K, V>(underlyingStore, journal, transactionContextFactory);
+        transactionManagerWrapper.registerXAResource(uniqueXAResourceId, xaResource);
+        transactionManagerWrapper.getTransactionManager().getTransaction().enlistResource(xaResource);
+        xaResources.put(transaction, xaResource);
```
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: Unless a transaction can only ever be bound to one thread, this should be a `putIfAbsent` guarding the synchronization registration.

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java [hunk 8]
```diff
@@ -0,0 +1,897 @@
...
+      advance();
+      return rc;
+    }
+  }
+
+  @Override
+  public ValueHolder<V> compute(K key, BiFunction<? super K, ? super V, ? extends V> mappingFunction, NullaryFunction<Boolean> replaceEqual) throws CacheAccessException {
+    checkKey(key);
+    XATransactionContext<K, V> currentContext = getCurrentContext();
+    if (currentContext.touched(key)) {
```
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: The whole if block can be extracted to a method `updateCommandForKey` to clarify flow and is duplicated in `computeIfPresent` so the method can be used there too.

**Comment by AbfrmBlr** [comment]
test this please

File: api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java (hunk 4 above)
Comment (ljacomet) [inline api/src/main/java/org/ehcache/exceptions/RethrowingCacheAccessException.java]: Thinking of it some more, no need for your impl to be synchronized since you call into super which will be (or not) depending on the VM running.

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java (hunk 7 above)
Comment (lorban) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: A transaction can only ever be bound/visible by one thread at a time, though it can migrate from one thread to another by the means of explicit calls to TransactionManager.suspend() / .resume(). And we create a new EhcacheXAResource instance for every transaction.

File: transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java (hunk 8 above)
Comment (lorban) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: Huh? AFAICT compute(), computeIfAbsent() and computeIfPresent() behave slightly differently in almost every aspect of their logic.
What am I overlooking?

Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java]: Inside the `currentContext.touched(key)` the code is the same. Maybe you need a condition before invoking the passed in function in `computeIfPresent`?

**Comment by AbfrmBlr** [comment]
test this please

**Comment by ljacomet** [comment]
test this please

**Comment by AbfrmBlr** [comment]
test this please
